lib/logflare_web/controllers/plugs/verify_api_access.ex

```diff
@@ -4,43 +4,51 @@ defmodule LogflareWeb.Plugs.VerifyApiAccess do
 
   Assigns the token's associated user if the token is provided
 
-  Options:
-  - `:resource`, required - an atom determining the resource accessed, as each resource has its own authorization verification flow.
+  Authentication api key can either be through access tokens or legacy `user.api_key`.
+
+  Access token usage is preferred and `user.api_key` is only used as a fallback.
   """
   import Plug.Conn
   import Phoenix.Controller
-  alias Logflare.{Auth, Endpoints}
+  alias Logflare.Auth
+  alias Logflare.Endpoints.Query
+  alias Logflare.Users
 
   def init(args), do: args |> Enum.into(%{})
 
-  def call(conn, opts) do
-    opts
-    |> Map.get(:resource)
-    |> do_auth(conn)
-  end
+  def call(%{assigns: %{endpoint: %Query{enable_auth: false}}} = conn, _opts), do: conn
 
-  defp do_auth(:endpoints, conn) do
-    conn = fetch_query_params(conn)
-    # fetch endpoint info
-    with endpoint_token <- conn.params["token"],
-         %Endpoints.Query{enable_auth: true, user_id: user_id} <-
-           Endpoints.get_query_by_token(endpoint_token),
-         {:ok, token} <- extract_token(conn),
-         {:ok, user} <- Auth.verify_access_token(token),
-         true <- user_id == user.id do
-      assign(conn, :user, user)
+  def call(conn, opts) do
+    opts = Enum.into(opts, %{scopes: []})
+    # generic access
+    with {:ok, user} <- identify_requestor(conn, opts.scopes) do
+      conn
+      |> assign(:user, user)
     else
-      %Endpoints.Query{enable_auth: false} ->
-        conn
-
       _ ->
         send_error_response(conn, 401, "Error: Unauthorized")
     end
   end
 
-  defp do_auth(_resource, conn) do
-    # unknown resource, reject as bad request
-    send_error_response(conn, 400, "Error: Bad request")
+  defp identify_requestor(conn, scopes) do
+    extracted = extract_token(conn)
+    is_private_route? = "private" in scopes
+
+    with {:ok, access_token_or_api_key} <- extracted,
+         {:ok, user} <- Auth.verify_access_token(access_token_or_api_key, scopes) do
+      {:ok, user}
+    else
+      {:error, :no_token} = err ->
+        err
+
+      {:error, _} = err ->
+        # try to use legacy api_key
+        case Users.get_by(api_key: elem(extracted, 1)) do
+          %_{} = user when is_private_route? == false -> {:ok, user}
+          _ when is_private_route? == true -> {:error, :unauthorized}
+          _ -> err
+        end
+    end
   end
 
```

Comment:
```suggestion
      {:error, _} = err -> handle_legacy_api_key(extracted, err)
    end
  end
  def handle_legacy_api_key(extracted, err) do
    case Users.get_by(api_key: elem(extracted, 1)) do
          %_{} = user when is_private_route? == false -> {:ok, user}
          _ when is_private_route? == true -> {:error, :unauthorized}
          _ -> err
    end
  end
```
This extracted to a function will make it easier to see that is an special case